realsense2_camera/src/realsense_node_factory.cpp

```diff
@@ -218,6 +304,83 @@ void RealSenseNodeFactory::StartDevice()
 	assert(_realSenseNode);
 	_realSenseNode->publishTopics();
 	_realSenseNode->registerDynamicReconfigCb(nh);
+
+	_initialized = true;
+
+	_data_timeout = privateNh.param("data_timeout", 0.0);
+	if (_data_timeout > 0.0)
+	{
+		ROS_INFO("starting data monitor ...");
+		_data_monitor_timer = nh.createTimer(ros::Duration(_data_timeout), &RealSenseNodeFactory::dataMonitor, this, false);
+	}
+	else
+	{
+		ROS_INFO("not monitoring data.");
+	}
+}
+
+bool RealSenseNodeFactory::shutdown()
+{
+	_initialized = false;
+	_data_monitor_timer.stop();
+
+	std::string manager_name = ros::this_node::getName();
+	std::string unload_service = manager_name + "/unload_nodelet";
+
+	if (ros::service::waitForService(unload_service, 0.1))
+	{
+		nodelet::NodeletUnload srv;
+		srv.request.name = getName();
+		if (!ros::service::call(unload_service, srv) || !srv.response.success)
+		{
+			ROS_WARN("Failed to unload nodelet, requesting shutdown ...");
+			ros::requestShutdown();
+		}
+	}
+	else
+	{
+		ROS_WARN("Failed to find unload nodelet service, requesting shutdown ...");
+		ros::requestShutdown();
+	}
+
+	return true;
+}
+
+bool RealSenseNodeFactory::reset()
+{
+	if (!_initialized)
+	{
+		return false;
+	}
+
+	_initialized = false;
+
+	_data_monitor_timer.stop();
+
+	_is_alive = false;
+	if (_query_thread.joinable())
+	{
+		_query_thread.join();
+	}
+	_realSenseNode.reset();
+	if (_device)
+	{
+		_device.hardware_reset();
```

Comment:
From what I read, when `reset()` is called, it performs `hardware_reset()` twice if the *rosparm* of `initial_reset` is set. This was not the case previously since the device enumeration was only performed once. Because of the abstraction of the `initialize()` method you have performed, the *rosparm* is re-read during initialization.

I recommend, resetting the *rosparm* itself (`initial_reset`) when its called for the first time [here](https://github.com/plusone-robotics/realsense/pull/28/files?file-filters%5B%5D=.h#diff-a7b463fc9149cf230a08dbb3e658221cR185). The reason I suggest against double reset is resetting takes time, and a double reset does not offer any benefit

Reply:
I modified the initialization to parse the `initial_reset` param in the `onInit()` function, which only gets called at the very beginning.  The `initial_reset` variable is set to false after initialization, so subsequent calls to the `reset` service won't result in a double reset.

Reply:
That seems to cut the time of the reset service from stopping publishing to starting again to ~ 3 seconds instead of ~13 seconds.

Reply:
Looks good now